File: cache/src/main/java/com/osrsd/cache/Library.java
package com.osrsd.cache;

import com.displee.cache.CacheLibrary;
import com.displee.cache.index.Index;
import com.displee.cache.index.archive.Archive;
import lombok.Synchronized;
import lombok.Value;
import lombok.extern.slf4j.Slf4j;

import java.util.Objects;

@Slf4j
@Value
public class Library {

    CacheLibrary cacheLibrary;

    public Library(final String path) {
        this.cacheLibrary = new CacheLibrary(path, false, (progress, message) -> log.info(String.format("%s and %s", progress, message)));
    }

    /**
     * Grabs a {@link Index} with the following id.
     * This is annotated with {@link Synchronized} due to some sort of problem with multiple threads accessing the library simultaneously.
     *
     * @param indexId The indexId to read from the cache library.
     * @return The index.
     */
    @Synchronized
    public Index index(final int indexId) {
        Index index = cacheLibrary.index(indexId);
        index.cache();
        return index;
    }

}

File: app/src/main/java/com/osrsd/cache/Library.kt
package com.osrsd.cache

import com.displee.cache.CacheLibrary
import com.displee.cache.index.Index
import com.osrsd.spring.App
import com.osrsd.spring.domain.*
import java.util.concurrent.ConcurrentHashMap

object Library {

    /**
     * The filesystem cache library.
     */
    private val cacheLibrary = CacheLibrary(App.baseDirectory())

    /**
     * Cached definitions provided from the cache library.
     */
    private val definitions: ConcurrentHashMap<Class<out Definition>, List<Definition>> = ConcurrentHashMap()

    /**
     * Gets the raw index from the cache library.
     * This is annotated with synchronized because of a weird issue with multiple threads accessing the library simultaneously.
     */
    @Synchronized
    fun index(indexId: Int): Index {
        val index: Index = cacheLibrary.index(indexId)
        index.cache()
        return index
    }

    /**
     * Stores a provided list of definitions.
     */
    fun store(clazz: Class<out Definition>, list: List<Definition>) {
        definitions[clazz] = list
    }

    /**
     * Gets the item definitions.
     */
    fun items(): List<ItemDefinition>? {
        return definitions[ItemDefinition::class.java]?.filterIsInstance<ItemDefinition>()
    }

    /**
     * Gets the npc definitions.
     */
    fun npcs(): List<NpcDefinition>? {
        return definitions[NpcDefinition::class.java]?.filterIsInstance<NpcDefinition>()
    }

    /**
     * Gets the enum definitions.
     */
    fun enums(): List<EnumDefinition>? {
        return definitions[EnumDefinition::class.java]?.filterIsInstance<EnumDefinition>()
    }

    /**
     * Gets the health bar definitions.
     */
    fun healthBars(): List<HealthBarDefinition>? {
        return definitions[HealthBarDefinition::class.java]?.filterIsInstance<HealthBarDefinition>()
    }

}